flowpipe/graph.py

```diff
@@ -108,62 +108,95 @@ def add_node(self, node):
             LogObserver.push_message(
                 "Node '{0}' is already part of this Graph".format(node.name))
 
-    def evaluate(self, threaded=False, submission_delay=0.1, raise_after=None):
-        """Evaluate all Nodes.
+    def evaluate(self, mode="linear", skip_clean=False,
+                 submission_delay=0.1, raise_after=None):
+        """Evaluate all Nodes in the graph.
+
+        Sorts the nodes in the graph into a resolution order and evaluates the
+        nodes. Nodes that do not depend on each other can be parallelized
+        either by threading or multiprocessing, see the "mode" keyword.
+
+        Note that no checks are in place whether the node execution is actually
+        thread-safe or fit for multiprocessing. It is assumed to be given if
+        the respective mode is selected.
 
         Args:
-            threaded (bool): Whether to execute each node in a separate thread.
+            mode (str): The evaluation mode. Possible modes are
+                * linear : Iterates over all nodes in a single thread
+                * threading : Spawns independent nodes in new threads
+                * multiprocessing : Spawns independent nodes in new processes
+            skip_clean (bool): Whether to skip nodes that are 'clean', i.e.
+                whose inputs have not changed since their output was computed
             submission_delay (float): The delay in seconds between loops
-                issuing new threads if nodes are ready to process.
+                issuing new threads/processes if nodes are ready to process.
             raise_after (int): The number of loops without currently running
-                threads after which to raise a RuntimeError.
+                threads/processes after which to raise a RuntimeError.
 
         """
         LogObserver.push_message("Evaluating Graph '{0}'".format(self.name))
-        if not threaded:
-            for node in self.evaluation_sequence:
+
+        eval_modes = {
+            "linear": self._evaluate_linear,
+            "threading": self._evaluate_threaded,
+            "multiprocessing": self._evaluate_multiprocessed
+        }
+
+        try:
+            eval_func = eval_modes[mode]
+        except KeyError:
+            mode_options = ""
+            for m in eval_modes:
+                mode_options += m + " "
+            mode_options = mode_options[:-1]  # get rid of trailing space
+            raise ValueError("Invalid mode {0}, options are {1}".format(
+                mode, mode_options))
+
+        eval_func(skip_clean=skip_clean, submission_delay=submission_delay,
+                  raise_after=raise_after)
+
+    def _evaluate_linear(self, skip_clean, **kwargs):
```

Comment:
I'd remove the kwargs since they're not used in the code. We can alwayas add new kwargs later and even change the signature altogether since this is a private method.

Reply:
If I drop the `kwargs` here, the nice factory pattern I use (having a dict of evaluation functions and calling the function we get from that), as `_evaluate_linear` will then complain about being handed a `submission_delay` etc.

Reply:
Ok, got it, then let's mention this in the docstring then to avoid future confusion